ENH: Allow specifying study, series, content datetime for DICOM image export

lassoan opened the pull request:
It was not possible to export 4D data sets, because study, series, frame of reference UIDs were always generated and it was not possible to set content time (to distinguish between image frames at the same spatial position at different time points).

Now study, series, frame of reference UIDs can be specified optionally and study, series, content date&time can be set.

cpinter commented:
What's the reason you fixed the old DICOM exporter module instead of doing something in the new DICOM exporter?

lassoan commented:
I am using the new exporter. I just needed improvements in the low-level infrastructure. The exporter plugin is here: https://github.com/SlicerRt/Sequences/commit/b9f29d13d7aa40287e86f01d552b7c1bfbe6de34

cpinter commented:
I see. Thanks for the explanation

fedorov commented on Modules/CLI/CreateDICOMSeries/CreateDICOMSeries.xml:
```diff
@@ -50,6 +50,13 @@
+    <string>
+      <name>studyTime</name>
+      <longflag>--studyTime</longflag>
+      <description><![CDATA[The time of the study [0008-0030]]]></description>
```
Is there a good reason to introduce a new notation conventions for DICOM tags here? Why not `(0008,0030)` or `(0x0008,0x0030)`? Same for "The time of the study" - why not `StudyTime` or `Study Time`?

lassoan replied:
I followed existing pattern. We can change all of them to standard style.

fedorov replied:
Makes sense. If it is existing pattern, I am fine leaving them as is.

fedorov commented on Modules/CLI/CreateDICOMSeries/CreateDICOMSeries.cxx:
```diff
@@ -222,6 +226,49 @@ int DoIt( int argc, char * argv[])
+      }
+    else if (keepOriginalUID)
+      {
+      std::cerr << "studyInstanceUID must be specified if any of the UIDs are specified." << std::endl;
```
Should there be a `return` here? If this is a warning, then maybe the message should not say "must"? Also, I do not understand the logic why SeriesInstanceUID would be required if StudyInstanceUID or FrameOfReferenceUID are specified? Can you comment?

lassoan replied:
If you prefer, we can return with error.

fedorov replied:
It is just confusing that the message is not consistent with the action. I would either return error, but probably better to just print a warning, because I don't see a good reason why this should be a "must".

lassoan replied:
> I don't see a good reason why this should be a "must"

[itkGDCMImageIO](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/IO/GDCM/src/itkGDCMImageIO.cxx#L1154-L1189) implements an all-or-nothing approach with these 3 UIDs. You either specify them or ITK generates all of them.

fedorov replied:
I see. Thanks for investigating. I don't understand that logic, but it is unrelated to this PR.

lassoan commented:
Some tags are pre-populated from DICOM import (if they are stored in Subject Hierarchy) but not all, and certainly SOP instance UID should match modality, so this exporter will only export simple time series correctly. It is a step forward and if someone has more time or specific need to export DCE-MRI, etc. correctly then there is a starting point.

fedorov commented:
Makes sense, I agree.

pieper commented:
👍 I agree we could always imagine a more perfect overall solution, but this is a step in the right direction.

lassoan commented:
Thanks for the reviews. I'll make the proposed changes and merge.

lassoan commented:
I've cleaned up the code a bit (changed DICOM tag to standard style, etc), added automatic SOP class UID setting based on image modality, and returning with error when not all UIDs are specified.

fedorov commented:
Thank you Andras, this is a good improvement!

fedorov commented on Modules/CLI/CreateDICOMSeries/CreateDICOMSeries.xml:
```diff
@@ -111,16 +132,55 @@
+    <string>
+      <name>contentDate</name>
+      <longflag>--contentDate</longflag>
+      <description><![CDATA[The date of the image content (0008,0023)]]></description>
```
Small nitpick - would be nice if "The date of the image content" and other similar occurrences were replaced with the actual attribute names from DICOM.

lassoan replied:
DICOM tag name is already in the "label" field (that is displayed in Create DICOM Series module GUI). This CLI module is just a backend anyway, user interacts with the DICOM module's GUI.

fedorov replied:
Yes, that is true. It is not critical.

lassoan commented:
This was committed a year ago in http://svn.slicer.org/Slicer4/trunk@27563.
